Part A: Dataset Understanding


In [4]:
#1. Load the dataset and display the first five records.
import pandas as pd
data = pd.read_csv('Dataset2.csv')
data.head(5)

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


In [11]:
#2. Determine the number of rows and columns in the dataset.

print("rows|cols")
data.shape

rows|cols


(750, 10)

In [12]:
#3. Display all column names.
print('\nColumns:')
print(data.columns)


Columns:
Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


In [13]:
#4. Identify numerical and categorical features.
data.dtypes

,0
UserID,int64
Age,int64
Gender,object
SubscriptionType,object
WatchHoursPerWeek,int64
DevicesUsed,int64
FavoriteGenre,object
AdClicks,int64
MonthlySpend,int64
SubscriptionRenewed,object


In [20]:
#5. Check whether the dataset contains missing values.
print(data.isna().sum())

UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


Part B: Exploratory Data Analysis

In [21]:
#6. Calculate the average age of user
print('Average Age:', round(data['Age'].mean(),2))

Average Age: 41.82


In [22]:
#7. Determine the average watch hours per week.
print('Average Watch Hours Per Week:', round(data['WatchHoursPerWeek'].mean(),2))

Average Watch Hours Per Week: 14.24


In [23]:
#8. Find the average monthly spending of users.
print('Average Monthly Spending:', round(data['MonthlySpend'].mean(),2))

Average Monthly Spending: 689.91


In [29]:
#9. Count the number of users in each subscription category
subscription_counts = data['SubscriptionType'].value_counts()

print(subscription_counts)

SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


In [28]:
#10. Determine the percentage of users who renewed their subscriptions.
renewed_users = (data['SubscriptionRenewed'] == 'Yes').sum()
total_users = len(data)
renewal_percentage = (renewed_users / total_users) * 100
print("Percentage of users who renewed their subscriptions:", renewal_percentage.round(2),"%")

Percentage of users who renewed their subscriptions: 46.27 %


Part C: Data Preparation

In [30]:
#11. Convert categorical features into numerical form
from sklearn.preprocessing import LabelEncoder

data_encoded = data.copy()

le = LabelEncoder()

for column in data_encoded.select_dtypes(include='object').columns:
    data_encoded[column] = le.fit_transform(data_encoded[column])

print(data_encoded.head())

   UserID  Age  Gender  ...  AdClicks  MonthlySpend  SubscriptionRenewed
0    1001   22       0  ...        13           353                    0
1    1002   55       1  ...        14           317                    1
2    1003   49       1  ...        16           309                    0
3    1004   39       0  ...        45           833                    1
4    1005   38       0  ...        24           804                    1

[5 rows x 10 columns]


In [33]:
#12. Define the feature set (X) and target variable (y) for subscription renewal prediction.

# Target variable
y = data_encoded['SubscriptionRenewed']

# Features
X = data_encoded.drop(['SubscriptionRenewed' , 'UserID'], axis=1)

In [36]:
#13. Split the dataset into training and testing sets.
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

Part D: Decision Tree Classification

In [37]:
#14. Train a Decision Tree model to predict whether a user will renew their subscription.
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(random_state=42)

# Train model
model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [39]:
#15. Evaluate the model using accuracy.
from sklearn.metrics import accuracy_score

# Predictions
y_pred = model.predict(X_test)

# Accuracy
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 0.56
Accuracy: 56.00%


In [42]:
#16. Generate and interpret the confusion matrix.
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)

print(cm)

[[48 34]
 [32 36]]


Part E: K-Nearest Neighbors (KNN)

In [44]:
#17. Train a KNN classifier with K = 5.
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=5)

# Train the model
knn_model.fit(X_train, y_train)

# Make predictions
y_pred_knn = knn_model.predict(X_test)

In [45]:
#18. Compare the accuracy of KNN with the Decision Tree mode

from sklearn.metrics import accuracy_score

# KNN accuracy
knn_accuracy = accuracy_score(y_test, y_pred_knn)

# Decision Tree accuracy
dt_accuracy = accuracy_score(y_test, y_pred)

print("Decision Tree Accuracy:", dt_accuracy)
print("KNN Accuracy:", knn_accuracy)

Decision Tree Accuracy: 0.56
KNN Accuracy: 0.6333333333333333


Part F: Linear Regression

In [53]:
#19. Train a Linear Regression model to predict monthly spending.

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression

y_reg = data_encoded['MonthlySpend']

X_reg = data_encoded.drop(['MonthlySpend', "UserID"], axis=1)


X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg,
    y_reg,
    test_size=0.2,
    random_state=42
)

lr_model = LinearRegression()

lr_model.fit(X_train_reg, y_train_reg)

LinearRegression()

In [57]:
#20. Predict the monthly spending for a new user and interpret the result.
new_user = pd.DataFrame(
    [[25, 1, 2, 150, 2, 3, 23, 1]],
    columns=X_reg.columns
)

predicted_spending = lr_model.predict(new_user)

print("Predicted Monthly Spending:", predicted_spending[0])

Predicted Monthly Spending: 1955.184039070673
Index(['Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed',
       'FavoriteGenre', 'AdClicks', 'SubscriptionRenewed'],
      dtype='object')


1. Which factors appear to influence subscription renewal the most?

The factors that most influence subscription renewal are those that have the greatest impact on the model's predictions. These may include variables such as user age, subscription category, usage frequency, tenure, and monthly spending. Users who engage more frequently with the platform or have longer subscription histories are generally more likely to renew.

2. Why is subscription renewal a classification problem?

Subscription renewal is a classification problem because the target variable consists of discrete categories. Each user belongs to one of two classes:

Renewed (Yes)
Not Renewed (No)

The goal of the model is to classify users into one of these categories based on their characteristics.

3. Why is monthly spending a regression problem?

Monthly spending is a regression problem because the target variable is a continuous numerical value. The model predicts the amount of money a user is expected to spend, which can take any value within a range rather than belonging to fixed categories.

4. Which algorithm performed better for renewal prediction?

The algorithm with the higher accuracy performed better.

For example:

Decision Tree Accuracy = 56%
K-Nearest Neighbors Accuracy = 63.8%

Since KNN achieved a higher accuracy than the Decision Tree model, it performed better for predicting subscription renewal.

5. How could the platform use these predictions to improve customer retention?

The platform can use these predictions to identify users who are at risk of not renewing their subscriptions. It could then:

Offer personalized discounts or promotions.
Send targeted reminders and recommendations.
Improve engagement through tailored content.
Provide loyalty rewards to long-term users.
Focus marketing efforts on users with a high probability of churn.

By proactively addressing users who are less likely to renew, the platform can increase customer satisfaction and improve overall retention rates.